src/database/accessors/loadProposalBundle.ts

```diff
@@ -1,16 +1,222 @@
-import { isProposal } from '../../types';
 import { loadBundle } from './loadBundle';
 import type { TinyPgParams } from 'tinypg';
 import type {
   Bundle,
   Proposal,
+  ProposalVersion,
+  ProposalFieldValue,
+  ApplicationFormField,
 } from '../../types';
 
+/* eslint-disable @typescript-eslint/no-type-alias */
+type AddPrefixToObject<T, P extends string> = {
+  [K in keyof T as K extends string ? `${P}${K}` : never]: T[K]
+};
+/* eslint-enable @typescript-eslint/no-type-alias */
+
+const filterByDistinctId = (
+  candidateItem: { id: number },
+  index: number,
+  self: { id: number }[],
+) => index === self.findIndex(
+  (item) => item.id === candidateItem.id,
+);
+
+// See https://github.com/typescript-eslint/typescript-eslint/issues/1824
+/* eslint-disable @typescript-eslint/indent */
+type ProposalPiece = AddPrefixToObject<Proposal, 'proposal_'>;
+type ProposalVersionPiece = AddPrefixToObject<ProposalVersion, 'proposalVersion_'>;
+type ProposalFieldValuePiece = AddPrefixToObject<ProposalFieldValue, 'proposalFieldValue_'>;
+type ApplicationFormFieldPiece = AddPrefixToObject<ApplicationFormField, 'applicationFormField_'>;
+/* eslint-enable @typescript-eslint/indent */
+
+const isApplicationFormFieldPiece = (
+  o: Record<string, unknown>,
+): o is ApplicationFormFieldPiece => (
+  'applicationFormField_id' in o
+);
+const isProposalFieldValuePiece = (
+  o: Record<string, unknown>,
+): o is ProposalFieldValuePiece => (
+  'proposalFieldValue_id' in o
+);
+const isProposalVersionPiece = (
+  o: Record<string, unknown>,
+): o is ProposalVersionPiece => (
+  'proposalVersion_id' in o
+);
+const isProposalPiece = (
+  o: Record<string, unknown>,
+): o is ProposalPiece => (
+  'proposal_id' in o
+);
+
+const extractApplicationFormFields = (
+  applicationFormFieldPieces: ApplicationFormFieldPiece[],
+): ApplicationFormField[] => (
+  applicationFormFieldPieces.reduce<ApplicationFormField[]>(
+    (applicationFormFields, applicationFormFieldPiece) => [
+      ...applicationFormFields,
+      {
+        id: applicationFormFieldPiece.applicationFormField_id,
```

Comment:
These extraction functions are unfortunately repeating the shape of the objects they are extracting to -- I don't know of any better way to do this, but it does mean we now have to update the type AND this extraction function.

Reply:
It took me so long to find where on this interface Ic ould actually respond to this comment!

My concern with this is that, as I understand it, it would mean baking implicit magic /into attribute names.  I've kind of made it a blanket rule for me that if I'm ever building business logic that depends on a naming pattern, I'm doing something that should be more explicit.

Basically: this approach would trade explicitness for a level of dynamic mapping that makes me a bit uncomfortable that we're being a bit too clever...

I'm also not entirely sure that the above would pass type safety muster, which means we could never further decorate or preflight-validate these objects

For instance, taking this approach would mean if we ever got fancy and figured out how to map our openAPI specification to our handlers, the handlers would not be returning objects that could be statically analyzed to being the valid / expected type.

Reply:
^  All this is to say...

I think if we go this route we do need to be explicit and verbose :-/

I definitely don't love that part of course, for the same reasons you don't -- (which honestly is why I had lobbied for that enrichment pattern way back when) -- but I can't quite think of a good solution that maintains explicitness without having to be redundant...

Maybe instead of passing <T> going back to passing a type guard for T would be enough -- that way we're confirming that the mapping is real and correct, and that we do indeed have type safety from our magically named fields.

Reply:
I just realized the json returns that are being explored [here](https://chat.opentechstrategies.com/#narrow/stream/75-PDC/topic/deep.20loading.20exploration/near/163035) might make this conversation moot to begin with!

Maybe we should hit pause on this for now until we resolve that?